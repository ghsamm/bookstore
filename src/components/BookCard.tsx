import React, { FunctionComponent } from 'react'
import styled from 'styled-components'
import Book from '../models/book'

interface Props {
    book: Book
}

interface ImageProps {
    link: string
}

const S = {
    BookCard: styled.div`
        border: 2px solid black;
        background-color: black;
        border-radius: 4px;
        overflow: hidden;
        aspect-ratio: 2 / 3;
        min-height: 250px;
        display: flex;
        flex-direction: column;
    `,
    Image: styled.div<ImageProps>`
        background-image: url(${props => props.link});
        background-position: center;
        background-size: cover;
        background-repeat: no-repeat;
        flex: 1;
    `,
    Content: styled.div`
        padding: 5px;
        background-color: black;
        color: white;
        display: flex;
        flex-direction: column;
        gap: 5px;
    `,
    Title: styled.div`
        text-align: center;
        font-weight: bold;
        font-size: 1.3em;
        line-height: 1.4;
    `,
    Author: styled.div`
        text-align: right;
        font-style: italic;
    `
}

const BookCard: FunctionComponent<Props> = ({ book }) => {
    return <S.BookCard>
        <S.Image link={book.cover} />
        <S.Content>
            <S.Title>{book.title}</S.Title>
            <S.Author>{book.author}</S.Author>
            <div></div>
        </S.Content>
    </S.BookCard>
}

export default BookCard