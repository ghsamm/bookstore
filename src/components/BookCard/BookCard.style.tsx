import { Link as ReactRouterLink } from 'react-router-dom'
import styled from 'styled-components'
import fromTheme from '../../style/fromTheme'
import BaseCard from '../BaseCard'
import BaseImage from '../BaseImage'

export const Link = styled(ReactRouterLink)`
    color: inherit;
    text-decoration: none;
`



export const Image = styled(BaseImage)`
    flex: 1;
`

export const Content = styled.div`
    padding: 5px;
    background-color: ${fromTheme(t => t.colors.main)};
    display: flex;
    flex-direction: column;
    gap: 5px;
`

export const Title = styled.div`
    text-align: center;
    font-weight: bold;
    font-size: 1.3em;
    line-height: 1.4;
`

export const Author = styled.div`
    text-align: right;
    font-style: italic;
`

export const SeeMore = styled.div.attrs({ className: 'see-more' })`
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-color: rgba(0,0,0,0.5);
    display: flex;
    align-items: center;
    justify-content: center;
    font-size: 1.2em;
    opacity: 0;
    transition: opacity 200ms ease-in-out;
`

export const BookCard = styled(BaseCard)`
    color: ${fromTheme(t => t.colors.textInverted)};
    aspect-ratio: 2 / 3;
    max-width: 300px;
    min-height: 250px;
    display: flex;
    flex-direction: column;
    position: relative;
    &:hover {
        .see-more {
            opacity: 1;
        }
    }
`