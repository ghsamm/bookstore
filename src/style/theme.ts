const theme = {
    colors: {
        main: 'black',
        secondary: 'grey',
        error: '#e36060',
        text: 'black',
        textInverted: 'white',
        label: 'grey'
    }
}

export type Theme = typeof theme

export default theme