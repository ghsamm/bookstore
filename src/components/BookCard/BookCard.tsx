import React, { FunctionComponent } from 'react'
import Book from '../../models/book'
import * as S from './BookCard.style'

interface Props {
    book: Book
}

const BookCard: FunctionComponent<Props> = ({ book }) => {


    return <S.Link to={`/books/${book.id}`}>
        <S.BookCard>
            <S.Image link={book.cover} />
            <S.Content>
                <S.Title>{book.title}</S.Title>
                <S.Author>{book.author}</S.Author>
            </S.Content>
            <S.SeeMore>See More Details</S.SeeMore>
        </S.BookCard>
    </S.Link>
}

export default BookCard