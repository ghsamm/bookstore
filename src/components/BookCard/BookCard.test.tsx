import React, { FunctionComponent } from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom'
import Book from '../../models/book';
import BookCard from './BookCard';
import { ThemeProvider } from 'styled-components';
import theme from '../../style/theme';


const Wrapper: FunctionComponent = ({ children }) => {
    return <MemoryRouter>
        <ThemeProvider theme={theme}>
            {children}
        </ThemeProvider>
    </MemoryRouter>
}

it('renders correctly', () => {

    const book: Partial<Book> = {
        id: '97978',
        cover: 'https://picsum.photos/640/480/?image=1000',
        title: 'Book of the year',
        author: 'Some one'
    }

    const tree = renderer
        .create(<Wrapper><BookCard book={book as Book} /></Wrapper>)
        .toJSON();
    expect(tree).toMatchSnapshot();
});