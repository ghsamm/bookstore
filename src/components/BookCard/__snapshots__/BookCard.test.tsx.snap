// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<a
  className="sc-bdfBwQ eDrhPS"
  href="/books/97978"
  onClick={[Function]}
>
  <div
    className="sc-gsTCUz sc-fubCfw bVmNIK hRzoSX"
  >
    <div
      className="sc-dlfnbm sc-hKgILt vRUoF bgpyTi"
    />
    <div
      className="sc-eCssSg fcoeFO"
    >
      <div
        className="sc-jSgupP kAcVGk"
      >
        Book of the year
      </div>
      <div
        className="sc-gKsewC jJKSsS"
      >
        Some one
      </div>
    </div>
    <div
      className="sc-iBPRYJ eKLUSM see-more"
    >
      See More Details
    </div>
  </div>
</a>
`;
